app/services/todo.js

```diff
@@ -1,8 +1,168 @@
+              axios.delete(`https://outlook.office.com/api/v2.0/me/tasks('${
+                taskId
+              }')`, {
+                headers: {"Authorization": `Bearer ${token}`},
+              }).then((res)=>{
+                resolve();
+              }).catch((err)=>{
+                if (err.response.status == 401) {
+                  this.requestRefresh().then(()=>{
+                    this.deleteTodo(taskId).then(resolve).catch(reject);
+                  }).catch((err)=>{
+                    console.error(err);
+                  });
+                } else {
+                  reject(err);
+                }
+              });
+            } else {
+              reject(new Error("ms_todo_folder_id is not saved"));
+            }
+          }).catch((err)=>{
+            reject(err);
+          });
+        } else {
+          reject(new Error("ms_todo_token is not saved"));
+        }
+      }).catch((err)=>{
+        reject(err);
+      });
+    });
+  };
+
+  this.requestRefresh = ()=>{
+    return new Promise((resolve, reject)=>{
+      preferences.get("ms_todo_refresh_token").then((rfToken)=>{
+        const queryParams = "client_id=" + process.env.MS_TODO_CLIENT_ID +
+        "&refresh_token=" + rfToken + "&client_secret=" + process.env.MS_TODO_CLIENT_SECRET +
+        "&grant_type=refresh_token";
+        axios.post("https://login.microsoftonline.com/common/oauth2/v2.0/token", queryParams,
+        ).then((tokenRes)=>{
+          const authToken = tokenRes.data.access_token;
+          preferences.set("ms_todo_token", authToken).then(()=>{
+            preferences.get("chat_id_ms_todo").then((chatId)=>{
+              resolve(authToken);
+            }).catch((err)=>{
+              reject(err);
+            });
+          }).catch((err)=>{
+          });
+        }).catch((err)=>{
+          reject(err);
+        });
+      }).catch((err)=>{
+        reject(err);
+      });
+    });
+  };
+  this.authorizeUser = (ctx) => {
+    preferences.set("chat_id_ms_todo", ctx.chat.id).then(()=>{
+      const base = "https://login.microsoftonline.com/common/oauth2/v2.0/authorize";
+      const client = `?client_id=${process.env.MS_TODO_CLIENT_ID}`;
+      const scope = "&scope=https%3A%2F%2Foutlook.office.com%2Ftasks.readwrite%20offline_access";
+      const responseType = "&response_type=code";
+      const redirectUri = "&redirect_uri=http://localhost:8080/mstodo";
+      ctx.reply("Bitte melde dich bei Microsoft Todo an:\n" + base + client + scope + responseType + redirectUri);
+    }).catch((err)=>{
+      ctx.reply("Tut mir Leid, da hat etwas nicht funktioniert");
+      console.error(err);
+    });
+  };
+
+  this.chooseFolder = (ctx, chatId) => {
+    preferences.get("ms_todo_token").then((token)=>{
+      axios.get("https://outlook.office.com/api/v2.0/me/taskfolders", {
+        headers: {"Authorization": `Bearer ${token}`},
+      }).then((res)=>{
+        const inlineKeyboardMarkup = {inline_keyboard: [[]]};
```

Comment:
sollte das nicht in den usecase eher?

Reply:
Nein, das ist unabhängig vom Usecase muss der user den ordner aussuchen, den der service watchen soll (evtl hast du neben einer uni liste ja noch eine Liste für Einkäufe, eine für haushaltssachen oder ne wishlist).

Reply:
ja aber du antwortest doch über telegramm. Damit ist es eine Nutzer interaktion und sollte von der Logik eher in den Use case meiner Meinung nach

Reply:
Ich glaube nicht, dass Nutzerinteraktion heißt, dass es ein usecase ist. Das ist teil des services und wird vom usecase nicht als usecase logik gebraucht, sondern der service braucht es einfach um zu funktionieren. Ganz unabhängig vom usecase, das ist einfach nur Zufall, dass nur mein Usecase das braucht. Das wäre also zusätzlcih auch gedoppelter code, wenn das jeder usecase selbst macht.